### PR TITLE
Rework IRawConversion changes in Revit and Common projects

### PR DESCRIPTION
**Pull request by AlanRynne**
Removes `IRawConversion<T>` and leaves just `IRawConversion<TIn,TOut>`

Re-arranges revit conversions to clarify which ones are `Raw` and which ones are "top level" `ToSpeckle` conversions.

Implicit usings to simplify imports

POC comments

### CODE REVIEW

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/LocationConversionToSpeckle.cs [hunk 1]
```diff
@@ -1,31 +1,33 @@
...
   }
 
-  public override Base RawConvert(DB.Location target)
+  public Base RawConvert(DB.Location target)
   {
     return target switch
     {
-      LocationCurve curve => _curveConverter.ConvertToBase(curve.Curve),
-      LocationPoint point => _xyzConverter.RawConvert(point.Point),
+      DB.LocationCurve curve => (_curveConverter.RawConvert(curve.Curve) as Base)!, // POC: ICurve and Base are not related but we know they must be, had to soft cast and then !.
```
Comment (BovineOx) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/LocationConversionToSpeckle.cs:29]: Okay me reaction is a bit :-O

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/BaseConversionToSpeckle.cs [hunk 2]
```diff
@@ -8,30 +7,21 @@ namespace Speckle.Converters.RevitShared.ToSpeckle;
...
-    IRawConversion<THost>,
     IRawConversion<THost, TSpeckle>
-  where TSpeckle : notnull
+  where TSpeckle : Base
 {
   public Base Convert(object target)
   {
-    return ConvertToBase((THost)target);
-  }
+    var result = RawConvert((THost)target);
```
Comment (BovineOx) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/BaseConversionToSpeckle.cs:15]: I don't suppose this hurts tbh

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/FamilyInstanceConversionToSpeckle.cs [hunk 3]
```diff
@@ -1,32 +1,32 @@
...
+    IRawConversion<DB.FamilyInstance, SOBR.RevitColumn> columnConverter
   )
   {
     _elementConverter = elementConverter;
     _beamConverter = beamConverter;
     _columnConverter = columnConverter;
   }
 
-  public override Base RawConvert(FamilyInstance target)
+  public override Base RawConvert(DB.FamilyInstance target)
```
Comment (BovineOx) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/FamilyInstanceConversionToSpeckle.cs:29]: So I suppose this is where having named IRawConversions would be handy and avoid the ever growing if statements/switch

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs [hunk 4]
```diff
@@ -47,11 +42,11 @@ ParameterObjectAssigner parameterObjectAssigner
     _parameterObjectAssigner = parameterObjectAssigner;
   }
 
-  public override RevitWall RawConvert(DB.Wall target)
+  public override SOBR.RevitWall RawConvert(DB.Wall target)
   {
-    RevitWall speckleWall = new() { family = target.WallType.FamilyName.ToString(), type = target.WallType.Name };
+    SOBR.RevitWall speckleWall = new() { family = target.WallType.FamilyName.ToString(), type = target.WallType.Name };
```
Comment (BovineOx) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs:47]: And I guess none of this is going to be null?

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs [hunk 5]
```diff
@@ -108,21 +104,21 @@ private void AssignHostedElements(Wall target, RevitWall speckleWall)
...
     }
   }
 
-  private List<CurveArray> GetWallVoids(Wall wall)
+  private List<DB.CurveArray> GetWallVoids(DB.Wall wall)
   {
-    List<CurveArray> curveArrays = new();
-    var profile = ((Sketch)_contextStack.Current.Document.Document.GetElement(wall.SketchId))?.Profile;
+    List<DB.CurveArray> curveArrays = new();
+    var profile = ((DB.Sketch)_contextStack.Current.Document.Document.GetElement(wall.SketchId))?.Profile;
```
Comment (BovineOx) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs:121]: Yeah must change this doc type in the context, but TBF I am thinking of taking it out of the context...

**Review by BovineOx — APPROVED** [review]
I added a bunch of comments but mostly they are questions..  I'd read them through see if anything should be changed before merging or POC comments adding/upgrading

File: DUI3-DX/Connectors/Revit/Speckle.Connectors.RevitShared/Plugin/RevitExternalApplication.cs [hunk 6]
```diff
@@ -74,7 +75,7 @@ private void _container_PreBuildEvent(object sender, ContainerBuilder containerB
     // tbe event can probably go
     // IRawConversions should be separately injectable (and not Require an IHostObject... or NameAndRank attribute)
     // Name and Rank can become ConversionRank or something and be optional (otherwise it is rank 0)
-    containerBuilder.RegisterRawConversions();
+    containerBuilder.RegisterRawConversions().InjectNamedTypes<IHostObjectToSpeckleConversion>();
```
[the remark this answers is not in the document]
Comment (AlanRynne) [reply DUI3-DX/Connectors/Revit/Speckle.Connectors.RevitShared/Plugin/RevitExternalApplication.cs:78]: RegisterRawConversions will register every type that implements `IRawConversion<Tin,TOut>` (including if a type implements it more than once).

Then we must also InjectNamedTypes for top-level conversions as we were doing before.

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/LocationConversionToSpeckle.cs (hunk 1 above)
Comment (AlanRynne) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/LocationConversionToSpeckle.cs:29]: This is horrible but it is coming from `Objects` and I didn't want to touch it, nor am I certain how we should fix it. Let's have a chat at some point regarding this.

We've been thinking about having `IBase` or `ISpeckleObject` to be the backing interface for `Base`, which would enable this but with some pretty big breaking changes.

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/BaseConversionToSpeckle.cs (hunk 2 above)
Comment (AlanRynne) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/BaseConversionToSpeckle.cs:15]: I'm still deciding on this one. I just removed the unnecessary method nesting.

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/FamilyInstanceConversionToSpeckle.cs (hunk 3 above)
Comment (AlanRynne) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/FamilyInstanceConversionToSpeckle.cs:29]: Yep! I think this is the next thing we should try to figure out in parallel, but for now I'm quite happy to do this bc it's simple, just as long as we flag them as POC

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs (hunk 5 above)
Comment (AlanRynne) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs:121]: Comment for this already exists in the contextSTack class.

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs (hunk 4 above)
Comment (AlanRynne) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs:47]: That's a very good question, I think it's not possible to have a wall without a family name and a wall type

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/ColumnConversionToSpeckle.cs [hunk 7]
```diff
@@ -1,18 +1,17 @@
...
 using Speckle.Converters.RevitShared.Helpers;
-using Line = Objects.Geometry.Line;
-using Point = Objects.Geometry.Point;
+using Speckle.Core.Models;
 
 namespace Speckle.Converters.RevitShared.ToSpeckle;
 
+// POC: There is no validation on this converter to prevent conversion from "not a Revit Beam" to a Speckle Beam.
+// This will definitely explode if we tried. Goes back to the `CanConvert` functionality conversation.
 public class ColumnConversionToSpeckle : IRawConversion<DB.FamilyInstance, SOBR.RevitColumn>
```
[the remark this answers is not in the document]
Comment (AlanRynne) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/ColumnConversionToSpeckle.cs:11]: In my mind, Raw conversions should always be consumed fully typed. If raw conversions are covariant/contravariant, then we can register them by the name of their first generic argument.

IRawConversion<Rhino.Point, Speckle.Point> can be assigned to `IRawConversion<Rhino.Point,Base>` or even return `object` if we only care about the input.

We also don't want this discovered directly by the converter, as multiple RawConversions take in `FamilyInstance` but only one ToHost conversion should be picked by the top-level converter.

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BeamConversionToSpeckle.cs [hunk 8]
```diff
@@ -1,23 +1,25 @@
...
 using Objects;
-using Objects.BuiltElements.Revit;
 using Speckle.Converters.Common;
 using Speckle.Converters.Common.Objects;
 using Speckle.Converters.RevitShared.Helpers;
+using Speckle.Core.Models;
 
 namespace Speckle.Converters.RevitShared.ToSpeckle;
 
+// POC: There is no validation on this converter to prevent conversion from "not a Revit Beam" to a Speckle Beam.
```
[the remark this answers is not in the document]
Comment (AlanRynne) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BeamConversionToSpeckle.cs:9]: We need to figure out how, Right now at least we should Throw if the input family instance is not the kind we were expecting. We don't want to convert columns as beams or viceversa.

**Pull request merged**